VPA: prune stale container aggregates, split recommendations over true number of containers

jkyros opened the pull request:
#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Previously we weren't cleaning up "stale" aggregates when container names changed (because of renames, removals) and that was resulting in: 
- VPAs showing recommendations for containers which no longer exist 
- Resources being split across containers which no longer exist (resulting in some containers ending up with resource limits too small for them to effectively live) 
- There was also a corner case where during a rollout after a container was renamed/removed from a deployment, we were counting the number of unique container names and not the actual number of containers in each pod, so we were splitting resources that shouldn't have been split. 

This PR is an attempt to clean up those stale aggregates without incurring too much overhead, and make sure that the resources get spread across the correct number of containers during a rollout. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #6744

#### Special notes for your reviewer:

There are probably a lot of different ways we can do the pruning of stale aggregates for missing containers:  
- I went with explicitly marking and sweeping them because it saved us an additional loop through all the pods and containers
- We could also just as easily just have a `PruneAggregates()` that runs after `LoadPods()` that goes through everything and removes them (or do this work as part of `LoadPods()` but that seems...deceptive?)
- We could probably also tweak the existing `garbageCollectAggregateCollectionStates` and run it immediately after `LoadPods()` every time but that might be expensive.

I'm not super-attached to any particular approach, I'd just like to fix this, so I can retool it if necessary.
- If I am being ignorant, and there are corner cases I'm missing, absolutely let me know 
- it probably need some tests/cleanup and I'll change the names of things to...whatever you want them to be. :smile: 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added pruning of container aggregates and changed container math so resources will no longer be split across the wrong number of containers
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jkyros committed:
VPA: fix erroneous container resource split

Previously we were dividing the resources per pod by the number of
container aggregates, but in a situation where we're doing a rollout and
the container names are changing (either a rename, or a removal) we're
splitting resources across the wrong number of containers, resulting in
smaller values than we should actually have.

This collects a count of containers in the model when the pods are
loaded, and uses the "high water mark value", so in the event we are
doing something like adding a container during a rollout, we favor the
pod that has the additional container.

There are probably better ways to do this plumbing, but this was my
initial attempt, and it does fix the issue.

jkyros committed:
VPA: clean orphaned container checkpoints

Previously we were only cleaning checkpoints after something happened to
the VPA or the targetRef, and so when a container got renamed the
checkpoint would stick around forever.

Since we're trying to clean up the aggregates immediately now, we need
to force the checkpoint garbage collection to clean up any checkpoints
that don't have matching aggregates.

If the checkpoints did get loaded back in after a restart,
PruneContainers() would take the aggregates back out, but we probably
shouldn't leave the checkpoints out there.

Signed-off-by: Max Cao <macao@redhat.com>

jkyros committed:
VPA: immediately prune stale vpa aggregates

Previously we were letting the rate limited garbage collector clean up
the aggregate states, and that works really well in most cases, but when
the list of containers in a pod changes, either due to the removal or
rename of a container, the aggregates for the old containers stick
around forever and cause problems.

To get around this, this marks all existing aggregates/initial
aggregates in the list for each VPA as "not under a VPA" every time
before we LoadPods(), and then LoadPods() will re-mark the aggregates as
"under a VPA" for all the ones that are still there, which lets us
easily prune the stale container aggregates that are still marked as
"not under a VPA" but are still wrongly in the VPA's list.

This does leave the ultimate garbage collection to the rate limited
garbage collector, which should be fine, we just needed the stale
entries to get removed from the per-VPA lists so they didn't affect VPA
behavior.

kwiesmueller commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -386,6 +406,55 @@ func (feeder *clusterStateFeeder) LoadVPAs() {
+	// TODO(jkyros): This only removes the container state from the VPA's aggregate states, there
+	// is still a reference to them in feeder.clusterState.aggregateStateMap, and those get
+	// garbage collected eventually by the rate limited aggregate garbage collector later.
+	// Maybe we should clean those up here too since we know which ones are stale?
```
Is it a lot of extra work to do that? Do you see any risks doing it here?

jkyros replied:
No, I don't think it's a lot of extra work, it should be reasonably cheap to clean them up here since it's just deletions from the other maps if the keys exist, I just didn't know all the history. 

It seemed possible at least that we were intentionally waiting to clean up the aggregates so if there was an unexpected hiccup we didn't just immediately blow away all that aggregate history we worked so hard to get? (Like maybe someone oopses, deletes their deployment, then puts it back? Right now we don't have to start over -- the pods come back in, find their container aggregates, and resume ?)

adrianmoisey replied:
Leaving it for the gc feels like the right option to me, for what it's worth

kwiesmueller commented on vertical-pod-autoscaler/pkg/recommender/model/cluster.go:
```diff
@@ -143,9 +143,28 @@ func (cluster *ClusterState) AddOrUpdatePod(podID PodID, newLabels labels.Set, p
+// the correct number and not just the number of aggregates that have *ever* been present. (We don't want minimum resources
+// to erroneously shrink, either)
+func (cluster *ClusterState) setVPAContainersPerPod(pod *PodState) {
+	for _, vpa := range cluster.Vpas {
```
I'm wondering if there is already a place where this logic could go so we don't have to loop over all VPAs for every pod again here.
In large clusters with a VPA to Pod ratio that's closer to 1 this could be a little wasteful.

jkyros replied:
Hmm, yeah, I struggled with finding a less expensive way without making too much of a mess. Unless I'm missing something (and I might be) we don't seem to have a VPA <--> Pod map -- probably because we didn't need one until now? At the very least I think I should gate this to only run if the number of containers in the pod is > 1. 

Like, I think our options are:

1. update the VPA as the pods roll through (which requires me to find the VPA for each pod like I did here) or
2. count the containers as we load the VPAs (but we load the VPAs _before_ we load the pods, so we'd have to go through the pods _again_, so that doesn't help us)
3. have the VPA actually track the pods it's managing, something like this: https://github.com/jkyros/autoscaler/commit/6ddc2085905fcb2dafb5397822b9f893d436af67 (could also just be an array of `PodID`s and we could look up the state so we could save the memory cost of the PodState pointer, but you know what I mean) 

I put it where I did (`option 1`)  because at least `LoadPods()` was already looping through all the pods so we could freeload off the "outer" pod loop and I figured we didn't want to spend the memory on `option 3`. If we'd entertain `option 3` and are okay with the memory usage, I can totally do that?

omerap12 replied:
I believe option 3 is the best approach. We can implement this in a follow-up PR.

sreber84 commented:
/remove-lifecycle rotten

adrianmoisey commented:
/ok-to-test
/assign

I want to see if I can help get this merged

maxcao13 committed:
VPA: Slightly improve runtime on splitting aggregates

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 committed:
VPA: Add e2e tests for pruning grace period and stale recommendations

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 commented:
Hi everyone, so John has take a hiatus and has left me with this PR, so after catching up, I guess we are still waiting for those conversations to resolve on which way do we want to go with those design decisions. The two commits I just put up are just a improvement on the existing implementation (assuming we will go with that, we don't have to), and some e2e tests to prove this works.

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -299,6 +305,20 @@ func (feeder *clusterStateFeeder) GarbageCollectCheckpoints() {
+			if !aggregateExists {
+				err = feeder.vpaCheckpointClient.VerticalPodAutoscalerCheckpoints(namespace).Delete(context.TODO(), checkpoint.Name, metav1.DeleteOptions{})
+				if err == nil {
+					klog.V(3).Infof("Orphaned VPA checkpoint cleanup - deleting %v/%v.", namespace, checkpoint.Name)
+				} else {
+					klog.Errorf("Cannot delete VPA checkpoint %v/%v. Reason: %+v", namespace, checkpoint.Name, err)
```
Could you switch to structured logging here? 
We've recently rolled out structured logging to the rest of the VPA

maxcao13 replied:
Thanks! Fixed in https://github.com/kubernetes/autoscaler/pull/6745/commits/01e0f4b47f596e78ca2dff804bc5896f21365f2d

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -396,6 +416,55 @@ func (feeder *clusterStateFeeder) LoadVPAs(ctx context.Context) {
+
+		for containerKey, container := range vpa.AggregateContainerStates() {
+			if !container.IsUnderVPA {
+				klog.V(4).Infof("Deleting Aggregate for VPA %s/%s: container %s no longer present", vpa.ID.Namespace, vpa.ID.VpaName, containerKey.ContainerName())
```
Could this be changed to structured logging?

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -396,6 +416,55 @@ func (feeder *clusterStateFeeder) LoadVPAs(ctx context.Context) {
+		}
+		for containerKey, container := range vpa.ContainersInitialAggregateState {
+			if !container.IsUnderVPA {
+				klog.V(4).Infof("Deleting Initial Aggregate for VPA %s/%s: container %s no longer present", vpa.ID.Namespace, vpa.ID.VpaName, containerKey)
```
Could this be changed to structured logging?

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -396,6 +416,55 @@ func (feeder *clusterStateFeeder) LoadVPAs(ctx context.Context) {
+		}
+	}
+	if initialAggregatesPruned > 0 || aggregatesPruned > 0 {
+		klog.Infof("Pruned %d aggregate and %d initial aggregate containers", aggregatesPruned, initialAggregatesPruned)
```
Could this be changed to structured logging?

adrianmoisey commented:
I thought of a scenario, which may be something that needs to be catered for.

if someone has a VPA on a CronJob with `successfulJobsHistoryLimit` set to 0, then the VPA could cleanup any recommendation that it has of that CronJob, since there are moments in time when no Pods exist.

I wonder if this can be fixed by:
1. Writing a tilmestep of when the Pod was last seen
2. Cleaning up only after a configurable time, and only if this cleanup feature is enabled.

maxcao13 commented:
> I thought of a scenario, which may be something that needs to be catered for.

> if someone has a VPA on a CronJob with successfulJobsHistoryLimit set to 0, then the VPA could cleanup any recommendation that it has of that CronJob, since there are moments in time when no Pods exist.

Sorry for the noise, and forgive me for not understanding, but what would be the problem here in terms of removing stale recommendations? Is it because the VPA is supposed to use identical recommendations for every container in a Job that gets created by a CronJob and we want to reuse the recommendations? If so, then apparently https://github.com/kubernetes/autoscaler/pull/6745#discussion_r1588278666 should solve this issue since the CronJob still exists. But I'm probably misunderstanding the question...

maxcao13 commented:
/remove-area cluster-autoscaler

adrianmoisey commented:
> > I thought of a scenario, which may be something that needs to be catered for.
> 
> > if someone has a VPA on a CronJob with successfulJobsHistoryLimit set to 0, then the VPA could cleanup any recommendation that it has of that CronJob, since there are moments in time when no Pods exist.
> 
> Sorry for the noise, and forgive me for not understanding, but what would be the problem here in terms of removing stale recommendations? Is it because the VPA is supposed to use identical recommendations for every container in a Job that gets created by a CronJob and we want to reuse the recommendations? If so, then apparently [#6745 (comment)](https://github.com/kubernetes/autoscaler/pull/6745#discussion_r1588278666) should solve this issue since the CronJob still exists. But I'm probably misunderstanding the question...

I think this is a possible scenario prior to this PR merging:
1. A CronJob and VPA are configured
2. On first run of the CronJob, VPA's recommendation is calculated and written to the VPA object
3. On second run of the CronJob, the admission-controller has a recommendation that it can apply to the new Pod
4. Step 3 repeats, and as the workload changes over time, so does the recommendation

But, I think this PR would change that scenario to the following:
1. A CronJob and VPA are configured
2. On first run of the CronJob, VPA's recommendation is calculated and written to the VPA object
3. VPA recommender cleans up the recommendation, as no Pods currently exist
4. On second run of the CronJob, no recommendation is applied, since no recommendation exists. But the resulting Pod will cause a recommendation to be created
5. VPA recommender cleans up the recommendation, as no Pods currently exist
6. On third run of the CronJob, no recommendation is applied, since no recommendation exists. But the resulting Pod will cause a recommendation to be created
7. Steps 5-6 repeat, effectively causing the CronJob to not have a VPA

This scenario is assuming that the CronJob is configured with `successfulJobsHistoryLimit` set to 0, and that the frequency of the recommendation cleanup is greater than the frequency of the CronJob.

maxcao13 commented:
> This scenario is assuming that the CronJob is configured with successfulJobsHistoryLimit set to 0, and that the frequency of the recommendation cleanup is greater than the frequency of the CronJob.

Got it, thanks for the detailed explanation!

This does happen and the recommendations are cleaned up before the next job gets scheduled, assuming the recommendation happens in time and the job schedule is long enough for the recommendation to prune the stale recommendations.

Thinking about your solution, I think it makes sense to add some sort of timeout feature for recommendations that refresh when they are used again -- pretty much what you mean by your cleanup feature. Maybe as a per VPA configuration in the VPA CR spec? Currently this is the current implemenation of whether recommendations will get recorded in the object: https://github.com/kubernetes/autoscaler/blob/c44d4f00334f2d76410c237d4675eba7c49bd3df/vertical-pod-autoscaler/pkg/recommender/routines/recommender.go#L107

So instead, the "matching pods" filtered pods should still include pods that have removed aggregates for a certain period of time based on a per VPA configuration, and maybe a global timeout as well?

Although thinking about it, there is a small window between the VPA deciding "we should remove this container recommendation, it doesn't exist anymore and we should stop splitting recommendations over it because it was a rename" vs "we shouldn't remove this container recommendation because its targetRef's top most controller is a cronjob so we should not remove the recommendation as jobs will quickly use it again." I think there would need to be a lot of care in setting this staleness timeout feature, and users would have to know how exactly how long they want the recommendation to linger for depending on if it's a cronjob, or a regular workload, for them to scale as they intended.

maxcao13 commented:
So maybe it is best to just look at if the topmost controller is CronJob and decide based on that to remove this extra complexity?

adrianmoisey commented:
> So maybe it is best to just look at if the topmost controller is CronJob and decide based on that to remove this extra complexity?

I was actually wondering about this option too.
It would also be done for all well known types. Rather than looking at the Pods to determine if a VPA recommendation, needs cleaning, look at the top most well known type.

maxcao13 commented:
> It would also be done for all well known types. Rather than looking at the Pods to determine if a VPA recommendation, needs cleaning, look at the top most well known type.

In this case, if we rename a container, and the deployment exists, wouldn't the original bug still happen and there would be stale recommendation for the old container? Maybe the aggregate would get deleted but there just might be a stale recommendation in the vpa object, I am understanding this right.

That's why I am wondering if we have to specially handle CronJobs since those containers that get created are supposed to be deleted.

maxcao13 committed:
VPA: Introduce pruningGracePeriod which allows a grace period before the updater prunes aggregateContainerStates/recommendations due to non-existent containers

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 commented:
Okay, in https://github.com/kubernetes/autoscaler/pull/6745/commits/5742d151d221f1843a690e0c546bbe8543c91369, I added `pruningGracePeriod` to per VPA `containerPolicies` so that people can specify graces periods for stale recommendations. Here's an example:
```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "*/3 * * * *"
  successfulJobsHistoryLimit: 0
  jobTemplate:
    spec:
      template:
        metadata:
          labels:
            app: hello
        spec:
          containers:
          - name: hello
            image: busybox:1.28
            imagePullPolicy: IfNotPresent
            command:
            - /bin/sh
            - -c
            - |
              tries=0
              while [ $tries -lt 30 ]; do
                echo "Hello, World $tries!"
                tries=$((tries+1))
                sleep 5
              done
          restartPolicy: OnFailure
---
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: hello
spec:
  targetRef:
      apiVersion: "apps/v1"
      kind:       "CronJob"
      name:       "hello"
  updatePolicy:
    updateMode: "Initial"
    minReplicas: 1
  resourcePolicy:
    containerPolicies:
    - containerName: "*"
      pruningGracePeriod: 3m
```
The grace period is long enough so that aggregates don't get pruned before the next job gets created and refreshes the last update time of the container, and so the recommendations still get preserved. If you specify a long grace period or the targetRef is CronJob, it will be pruned in 24 hours. If you don't specify the field and the targetRef is not CronJob, then the `pruningGracePeriod` is 0 and aggregates will be pruned immediately as soon as the next recommendation loop occurs, which will then remove the recommendations from the api object.

IDK how people feel about this, so I didn't write any tests. But using it locally, it seems to work as intended.

adrianmoisey commented:
> > It would also be done for all well known types. Rather than looking at the Pods to determine if a VPA recommendation, needs cleaning, look at the top most well known type.
> 
> In this case if we do this, if we rename a container, and the deployment exists, wouldn't there still be stale recommendations for the old container? Maybe the aggregate would get deleted and we wouldn't split resources anymore, but there just might be a stale recommendation in the vpa object that will never get deleted (but maybe that's okay?), if I am understanding this right.

My idea was to check the targetRef to see what it contains, in order to determine if a VPA's recommendation needs cleaning.
Ie: for a Deployment targetRef, specifically looking at the list of `Deployment.spec.template.spec.containers[*].name` to figure out which containers recommendations to keep.

This will only work on well-known types

adrianmoisey commented:
> Okay, in [b6e435b](https://github.com/kubernetes/autoscaler/commit/b6e435bebde2a295229dac410577942da082ed9f), I added `pruningGracePeriod` to per VPA `containerPolicies` so that people can specify graces periods for stale recommendations.

My opinion is that this is an adequate solution.

@voelzmo @raywainman @kwiesmueller @omerap12 thoughts on this one?

omerap12 commented:
> > Okay, in [b6e435b](https://github.com/kubernetes/autoscaler/commit/b6e435bebde2a295229dac410577942da082ed9f), I added `pruningGracePeriod` to per VPA `containerPolicies` so that people can specify graces periods for stale recommendations.
> 
> My opinion is that this is an adequate solution.
> 
> @voelzmo @raywainman @kwiesmueller @omerap12 thoughts on this one?

I’m not sure I understand the use case for allowing users to specify grace periods for stale recommendations.
For deployments where container names have changed, couldn’t we use a default value or immediately discard stale recommendations?
For CronJobs, why not calculate the value dynamically based on the schedule? For example, if the schedule is 1 hour, we could set it to 1 hour + 10 minutes as a safeguard.

What do you think?

omerap12 commented on vertical-pod-autoscaler/pkg/utils/vpa/api.go:
```diff
@@ -218,6 +219,22 @@ func GetContainerControlledValues(name string, vpaResourcePolicy *vpa_types.PodR
+		if targetRef != nil && targetRef.Kind == "CronJob" {
+			// CronJob is a special case, because they create containers they are usually supposed to be deleted after the job is done.
+			// So we set a higher default grace period so that future recommendations for the same workload are not pruned too early.
+			// TODO(maxcao13): maybe it makes sense to set the default based on the cron schedule?
```
IMHO It does make sense.

adrianmoisey commented:
> If you don't specify the field and the targetRef is not CronJob, then the `pruningGracePeriod` is 0 and aggregates will be pruned immediately as soon as the next recommendation loop occurs, which will then remove the recommendations from the api object.

I think I figured out another edge-case that needs to be handled.

Sometimes in my day-job, when we're firefighting issues, we may scale a deployment down to zero to alleviate pressure on other parts of the system. I think this use-case is possibly common, tools such as https://keda.sh are built to allow users to scale workloads down to zero, to save costs.

During these times the recommendation would be removed. I think the grace period needs to apply to all resource types.

maxcao13 commented:
> I’m not sure I understand the use case for allowing users to specify grace periods for stale recommendations. For deployments where container names have changed, couldn’t we use a default value or immediately discard stale recommendations? For CronJobs, why not calculate the value dynamically based on the schedule? For example, if the schedule is 1 hour, we could set it to 1 hour + 10 minutes as a safeguard.
> 
> What do you think?

I think allowing the users to set it will be good in a case like @adrianmoisey said, where you may scale a deployment down to 0 to alleviate pressure on the system but don't want the recommendations to get lost. Then you can specifically set the `pruningGracePeriod` on the deployment's containers to be something high like `10d`, so that the recommendations don't get removed while you are troubleshooting your system. But maybe you don't want that for some other workload, so you can set the gracePeriod to something like 10 minutes, so we don't have a memory leak in the recommender.

> I think I figured out another edge-case that needs to be handled.

> Sometimes in my day-job, when we're firefighting issues, we may scale a deployment down to zero to alleviate pressure on other parts of the system. I think this use-case is possibly common, tools such as https://keda.sh/ are built to allow users to scale workloads down to zero, to save costs.

> During these times the recommendation would be removed. I think the grace period needs to apply to all resource types.

This _should_ be able to work on all types as long as you set the field (unless I am missing something). Do you mean that we should default the gracePeriod to be very high (never expire) instead of 0? Maybe we also allow a global default gracePeriod set with a command line flag?

adrianmoisey commented:
> > I think I figured out another edge-case that needs to be handled.
> 
> > Sometimes in my day-job, when we're firefighting issues, we may scale a deployment down to zero to alleviate pressure on other parts of the system. I think this use-case is possibly common, tools such as https://keda.sh/ are built to allow users to scale workloads down to zero, to save costs.
> 
> > During these times the recommendation would be removed. I think the grace period needs to apply to all resource types.
> 
> This _should_ be able to work on all types as long as you set the field (unless I am misunderstanding you). Do you mean that we should default the gracePeriod to be very high (never expire) instead of 0? Maybe we also allow a global default gracePeriod set with a command line flag?

Yeah, I guess it makes sense to have a long default or keep the feature disabled and allow it to be opt-in.

raywainman commented:
Could the default be "don't clean them up" to match behavior today? I worry that this is verging on a breaking change.

Then a user can override the pruning grace period for the problematic VPAs that are churning containers?

omerap12 commented:
> > > I think I figured out another edge-case that needs to be handled.
> > 
> > 
> > > Sometimes in my day-job, when we're firefighting issues, we may scale a deployment down to zero to alleviate pressure on other parts of the system. I think this use-case is possibly common, tools such as https://keda.sh/ are built to allow users to scale workloads down to zero, to save costs.
> > 
> > 
> > > During these times the recommendation would be removed. I think the grace period needs to apply to all resource types.
> > 
> > 
> > This _should_ be able to work on all types as long as you set the field (unless I am misunderstanding you). Do you mean that we should default the gracePeriod to be very high (never expire) instead of 0? Maybe we also allow a global default gracePeriod set with a command line flag?
> 
> Yeah, I guess it makes sense to have a long default or keep the feature disabled and allow it to be opt-in.

I agree

maxcao13 committed:
VPA: Add globalPruningGracePeriod flag which defaults to a long time for non-breaking opt-in change

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 commented:
Okay, I've added a `--pruning-grace-period-duration` flag which defaults to 100 years meaning that aggregates will not expire which is the previous implementation before this PR (hopefully someone is not keeping a vpa-recommender container up for 100 years).

I removed any special handling of CronJob since the default pruning functionality should now be the same as before.

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go:
```diff
@@ -46,6 +47,10 @@ import (
 )
 
+var (
+	globalPruningGracePeriodDuration = flag.Duration("pruning-grace-period-duration", 24*365*100*time.Hour, `The grace period for deleting stale aggregates and recommendations. By default, set to 100 years, effectively disabling it.`)
```
I'm unsure what others think, but I'm not super excited by having the default be 100 years. My preference would be that setting the global to zero would disable the pruning feature (except for VPAs with pruningGracePeriod set)

maxcao13 replied:
Yeah, I agree it's not ideal. I was trying to figure out how I could enable/disable it with a single flag, and without specially handling a zero value for a duration. I guess I could switch to `flag.String` and specially handle a `0` as off. Or if there's some cleaner way to do it I'm happy for feedback.

adrianmoisey replied:
For what it's worth, I found another flag that uses `<= 0` as a way to disable it: https://github.com/kubernetes/autoscaler/blob/ce01f0210f2b0d50f05395c7968752f548c8a75c/vertical-pod-autoscaler/pkg/updater/main.go#L62-L64

maxcao13 replied:
I changed to specially handle a `flag.String` in https://github.com/kubernetes/autoscaler/pull/6745/commits/72ae5d7e3908498eda829891c529533f94dfcf34

adrianmoisey commented:
Generally this seems OK to me.
I just want to give it a test drive locally before giving it a lgtm

I'd also like other approvers to weigh in here too

adrianmoisey commented on vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go:
```diff
@@ -215,6 +215,12 @@ type ContainerResourcePolicy struct {
+	// By default, recommendations for non-existent containers are never pruned until its top-most controller is deleted,
+	// after which the recommendations are subject to the VPA's recommendation garbage collector.
+	// +optional
+	PruningGracePeriod *metav1.Duration `json:"pruningGracePeriod,omitempty" protobuf:"bytes,4,opt,name=pruningGracePeriod"`
```
I'm unsure how others feel about this, but this change adds PruningGracePeriod to `verticalpodautoscaler.spec.resourcePolicy.containerPolicies`

Where the description of containerPolicies is:
>    Per-container resource policies.
    ContainerResourcePolicy controls how autoscaler computes the recommended
    resources for a specific container.


Technically speaking, PruningGracePeriod isn't related to how the VPA generates recommendations. It feels wrong to put it there, but I don't know of a better location to put it.

An idea I have, which I'm not super excited by, is to use an annotation on the VPA object to drive this.

omerap12 replied:
I feel the flags are a bit messy. Some are in the VPA object, some are global in the recommender, and some are in annotations. I think most flags related to how the VPA generates recommendations should be in the VPA object, while others should go in annotations. So yes, I agree.

maxcao13 replied:
Sure, makes sense to me. Thanks!

maxcao13 replied:
Changed in https://github.com/kubernetes/autoscaler/pull/6745/commits/70826a079bf5dc496743757cfacafeb63b3da3d9

maxcao13 committed:
VPA: refactor pruning grace period as vpa apiObject annotation instead of containerPolicy

Signed-off-by: Max Cao <macao@redhat.com>

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go:
```diff
@@ -46,6 +47,22 @@ import (
+	}
+	duration, err := time.ParseDuration(*globalPruningGracePeriodDuration)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to parse --pruning-grace-period-duration: %v", err))
```
Instead of using `panic()`, we should return an error to be handled by the caller of this function. Could we modify this to return an error using `klog.ErrorS()` followed by `os.Exit(255)`?
What do you think?

maxcao13 replied:
Fixed in https://github.com/kubernetes/autoscaler/pull/6745/commits/b2ae65af20c3bb75c5475deecf1241f523946531, does it make sense? I used `klog.Fatalf` instead of error and exit.

omerap12 replied:
Yes, that's fine, but could you use `Errors` and `Exit`? It's the correct approach for structural logging.

maxcao13 replied:
Hopefully https://github.com/kubernetes/autoscaler/pull/6745/commits/cd63322b24f38c9f5a3b896d5c6bf03c062f1111 is correct? :crossed_fingers:

omerap12 replied:
Yup! Thanks

adrianmoisey commented:
There's potentially something wrong here. I tested it locally. 
I made a deploy with 2 containers, and set the annotation `vpaPruningGracePeriod: 70s` on the VPA (may be this value is too short? I'm unsure).

Everything was fine, I was getting recommendations as expected.

I then deleted the second container, and eventually saw this in the logs:
```
I0105 13:04:21.715400       1 cluster_feeder.go:448] "Deleting Aggregate for VPA: container no longer present" namespace="default" vpaName="hamster-vpa" containerName="hamster"
I0105 13:04:21.715442       1 cluster_feeder.go:448] "Deleting Aggregate for VPA: container no longer present" namespace="default" vpaName="hamster-vpa" containerName="hamster2"
```

`hamster2` was the only container I had removed, but for some reason it removed both.

A while later I re-added hamster2, and eventually the VPA did this:
```
I0105 13:10:21.726150       1 cluster_feeder.go:448] "Deleting Aggregate for VPA: container no longer present" namespace="default" vpaName="hamster-vpa" containerName="hamster"
```

My guess is that it's related to the `MarkAggregates()` run, but I'm not too sure.

maxcao13 commented:
> There's potentially something wrong here. I tested it locally. I made a deploy with 2 containers, and set the annotation `vpaPruningGracePeriod: 70s` on the VPA (may be this value is too short? I'm unsure).
> 
> Everything was fine, I was getting recommendations as expected.
> 
> I then deleted the second container, and eventually saw this in the logs:
...
> My guess is that it's related to the `MarkAggregates()` run, but I'm not too sure.

What's happening is that removing the container makes the deployment deploy a new pod, and the VPA doesn't associate both of the old container aggregates with the new pod, so it never marks `isUnderVPA=true` for the next `LoadPods` loop. From what I can tell, this is probably okay because `AggregateStateKey` includes info about the pod's labelSet to make them unique.

adrianmoisey commented:
> > There's potentially something wrong here. I tested it locally. I made a deploy with 2 containers, and set the annotation `vpaPruningGracePeriod: 70s` on the VPA (may be this value is too short? I'm unsure).
> > Everything was fine, I was getting recommendations as expected.
> > I then deleted the second container, and eventually saw this in the logs:
> > ...
> > My guess is that it's related to the `MarkAggregates()` run, but I'm not too sure.
> 
> What's happening (I think) is that removing the container makes the deployment deploy a new pod, and the VPA doesn't associate both of the old container aggregates with the new pod, so it never marks `isUnderVPA=true` for the next `LoadPods` loop.
> 
> From what I can tell, this is probably okay because `AggregateStateKey` includes info about the pod's labelSet to make them unique (is there a situation where this breaks?). We don't want the old pod's aggregate to contribute to the recommendation anymore (assuming that's the reason someone sets the pruning grace period).

That theory sounds plausible. I worry that it's confusing to the user though, since it sounds like their recommendation gets removed.

What I'm mostly worried about is this part:

> A while later I re-added hamster2, and eventually the VPA did this:

> I0105 13:10:21.726150       1 cluster_feeder.go:448] "Deleting Aggregate for VPA: container no longer present" namespace="default" vpaName="hamster-vpa" containerName="hamster"
My guess is that it's related to the MarkAggregates() run, but I'm not too sure.

When this happened, the recommendations were deleted, and the status of my VPA looked like this:

```
status:
  conditions:
  - lastTransitionTime: "2025-01-07T19:50:10Z"
    status: "False"
    type: RecommendationProvided
  recommendation: {}
```

maxcao13 committed:
VPA: fix setting VPAContainersPerPod correctly; allow old aggregates to be linked to new containers

We need to set VPAContainersPerPod for a VPA correctly so we can split resources correctly on its first run through the recommendation loop. So I opted to explicitly set it after updating the pod's containers.

This also allows old aggregateContainerStates that were previously created from a removed Pod's container, to be reused by a new Pod's container that shares the same VPA and targetRef. This allows recommendations to be updated correctly when aggregates are pruned or created.

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 committed:
VPA: resolve various TODOs and review comments

Also sets the VPAContainersPerPod correctly when initializing VPA from history provider.

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 commented:
> That theory sounds plausible. I worry that it's confusing to the user though, since it sounds like their recommendation gets removed.
> 
> What I'm mostly worried about is this part:
> 
> > A while later I re-added hamster2, and eventually the VPA did this:
> 
> > I0105 13:10:21.726150       1 cluster_feeder.go:448] "Deleting Aggregate for VPA: container no longer present" namespace="default" vpaName="hamster-vpa" containerName="hamster"
> > My guess is that it's related to the MarkAggregates() run, but I'm not too sure.
> 
> When this happened, the recommendations were deleted, and the status of my VPA looked like this:
> 
> ```
> status:
>   conditions:
>   - lastTransitionTime: "2025-01-07T19:50:10Z"
>     status: "False"
>     type: RecommendationProvided
>   recommendation: {}
> ```

I wasn't able to make the bug appear that removes the recommendations completely, but I noticed that the pruning stale aggregates wouldn't update the number of container recommendations to the new number of pod containers, which is probably the same bug. 

I fixed this in https://github.com/kubernetes/autoscaler/pull/6745/commits/11bcee3a2a8cdf071d0823300fe0213ebecc750c hopefully. I needed to link old aggregates that were used by a previous pod/container, to the new pod/container explicitly. This is because the `cluster.findOrCreateAggregateContainerState(containerID)` function will not create a new aggregate. It finds the old one, and never calls `UseAggregationIfMatching` which never links this aggregation to the VPA object if the old aggregates are pruned.  Then the `GetContainerNameToAggregateStateMap` function would be able to _aggregate_ all the `containerAggregateStates` (super confusing naming) by container name based on how many non-stale aggregates actually still exist for the recommender post processor.

adrianmoisey commented:
> I wasn't able to make the bug appear that removes the recommendations completely, but I noticed that the pruning stale aggregates wouldn't update the number of container recommendations to the new number of pod containers, which is probably the same bug.

Yeah, it seems to be fixed now.

Something I noticed, was that the old recommendation still exists.

When I have 1 containers per Pod:
```
status:
  conditions:
  - lastTransitionTime: "2025-01-08T11:56:36Z"
    status: "True"
    type: RecommendationProvided
  recommendation:
    containerRecommendations:
    - containerName: hamster
      lowerBound:
        cpu: 527m
        memory: 131072k
      target:
        cpu: 627m
        memory: 131072k
      uncappedTarget:
        cpu: 627m
        memory: 131072k
      upperBound:
        cpu: "1"
        memory: 500Mi
    - containerName: hamster2
      lowerBound:
        cpu: 537m
        memory: 131072k
      target:
        cpu: 627m
        memory: 131072k
      uncappedTarget:
        cpu: 627m
        memory: 131072k
      upperBound:
        cpu: "1"
        memory: 500Mi
```

(notice that the memory is half of the default value of `--pod-recommendation-min-memory-mb`, as expected).

Then when I remove a container from the Deployment, I get this:
```
status:
  conditions:
  - lastTransitionTime: "2025-01-08T11:56:36Z"
    status: "True"
    type: RecommendationProvided
  recommendation:
    containerRecommendations:
    - containerName: hamster
      lowerBound:
        cpu: 558m
        memory: 262144k
      target:
        cpu: 627m
        memory: 262144k
      uncappedTarget:
        cpu: 627m
        memory: 262144k
      upperBound:
        cpu: "1"
        memory: 500Mi
    - containerName: hamster2
      lowerBound:
        cpu: 210m
        memory: 262144k
      target:
        cpu: 627m
        memory: 262144k
      uncappedTarget:
        cpu: 627m
        memory: 262144k
      upperBound:
        cpu: "1"
        memory: 500Mi
```

Notice that the second (now removed) container persists, and the memory (for both containers) is now correct.

I had assumed that as part of this PR, the removed recommendation should be removed from the status field.

maxcao13 commented:
I assume for the VPA, there is no pruningGracePeriod set or it's very long? I get the same thing when I do that.

If so, yeah, I guess that's a side effect of the recommendations not having getting pruned - that there will be stale recommendations still in the VPA. To make it less confusing to the user, maybe there could be an extra field to mark it as visibly stale? Maybe it would make more sense that the recommendation for `hamster2` is not updated at all if it goes away?
e.g.
`vpaPruningGracePeriod not set; container "hamster2" was removed"`
```
    recommendation:
      containerRecommendations:
      - containerName: hamster
        lowerBound:
          cpu: 582m
          memory: 262144k
        target:
          cpu: 587m
          memory: 262144k
        uncappedTarget:
          cpu: 587m
          memory: 262144k
        upperBound:
          cpu: "1"
          memory: 262144k
      - containerName: hamster2
        lowerBound:
          cpu: 403m
          memory: same as before
        target:
          cpu: 587m
          memory: same as before
        uncappedTarget:
          cpu: 587m
          memory: same as before
        upperBound:
          cpu: "1"
          memory: 500Mi
        stale: true
```

If you set a really short pruning period and that still happens, it shouldn't happen and this would be a bug (hoping it's not this case).

adrianmoisey commented:
Oh hold on, I was too impatient and wasn't waiting for the grace period to actually remove recommendation. My bad!

> Maybe it would also make more sense that the recommendation for hamster2 is not updated at all if it goes away?

I think this may make sense, updating the recommendation for a container that's about to be removed seems to show a sign of life, may be it needs to be ignored instead.

> The issue here, would still be covered as if there are no pods for a deployment, we don't remove anything, and the cronJob issue would be solved as well. Is there any bugs I'm missing in this solution?

Nope, I think the solution is fine at the moment (besides the point above, which is a mild confusion, and not a big deal), I was just not waiting the grace period that I had set.

I'll give this a few more tests though, just to see if anything else crops up

maxcao13 commented:
Hm... ignore the latest commit, I think I messed something up.

maxcao13 commented:
> >Maybe it would also make more sense that the recommendation for hamster2 is not updated at all if it goes away?

> I think this may make sense, updating the recommendation for a container that's about to be removed seems to show a sign of life, may be it needs to be ignored instead.

Latest commit should fix this now.

adrianmoisey commented:
I have another comment, and I'm unsure on this one.

The new behaviour results in this:
```
status:
  conditions:
  - lastTransitionTime: "2025-01-11T12:07:41Z"
    status: "True"
    type: RecommendationProvided
  recommendation:
    containerRecommendations:
    - containerName: hamster
      lowerBound:
        cpu: 445m
        memory: 262144k
      target:
        cpu: 627m
        memory: 262144k
      uncappedTarget:
        cpu: 627m
        memory: 262144k
      upperBound:
        cpu: "1"
        memory: 500Mi
    - containerName: hamster2
      lowerBound:
        cpu: 214m
        memory: 131072k
      target:
        cpu: 627m
        memory: 131072k
      uncappedTarget:
        cpu: 627m
        memory: 131072k
      upperBound:
        cpu: "1"
        memory: 500Mi
```

This happens after the second container is removed, but before the `vpaPruningGracePeriod` period happens.
The remaining container's memory is set to the value of `--pod-recommendation-min-memory-mb`, and the deleted container still exists.

What I had originally expected to happen was that the `--pod-recommendation-min-memory-mb` would be split between the two containers, until the `vpaPruningGracePeriod` period expires, then the stale container would be removed, and `--pod-recommendation-min-memory-mb` be recalculated.

But, thinking it over, I guess this is fine? The second container's recommendations are retained (until it comes back, or when `vpaPruningGracePeriod` elapses). The primary container gets recommendations that make sense.

adrianmoisey commented:
Generally speaking I'm fine with this change. I haven't given it a thorough review, but it works locally in a way that makes sense.
I see there's a TODO item to add more tests, and documentation still needs updating.

But I kinda want to get more thoughts on this (ping @omerap12 @raywainman @voelzmo) just to check if it all makes sense to someone else too

omerap12 commented:
> Generally speaking I'm fine with this change. I haven't given it a thorough review, but it works locally in a way that makes sense. I see there's a TODO item to add more tests, and documentation still needs updating.
> 
> But I kinda want to get more thoughts on this (ping @omerap12 @raywainman @voelzmo) just to check if it all makes sense to someone else too

Thanks for pinging, Ill take a look tomorrow

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/model/cluster.go:
```diff
@@ -130,29 +130,41 @@ func (cluster *ClusterState) AddOrUpdatePod(podID PodID, newLabels labels.Set, p
-func (cluster *ClusterState) addPodToItsVpa(pod *PodState) {
+// SetVPAContainersPerPod sets the number of containers per pod seen for pods connected to this VPA
+// so that later when we're splitting the minimum recommendations over containers,  we're splitting them over
+// the correct number and not just the number of aggregates that have *ever* been present. (We don't want minimum resources
```
🚀

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/model/cluster.go:
```diff
@@ -259,22 +282,42 @@ func (cluster *ClusterState) AddOrUpdateVpa(apiObject *vpa_types.VerticalPodAuto
-		vpaExists = false
 	}
 	if !vpaExists {
 		vpa = NewVpa(vpaID, selector, apiObject.CreationTimestamp.Time)
```
Can we log that we are creating a new VPA?

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender.go:
```diff
@@ -100,6 +102,26 @@ func (r *recommender) UpdateVPAs() {
+					containerName := resources.ContainerName
+					containerAggregateState := vpaContainerNameToAggregateStateMap[containerName]
+					if containerAggregateState != nil && !containerAggregateState.IsUnderVPA && !containerAggregateState.IsAggregateStale(now) {
+						klog.V(5).InfoS("Container no longer exists, but is not stale. Keeping container recommendation at previous state.", "vpa", vpa.ID.VpaName, "container", containerName)
```
I'm not sure I follow. If the container no longer exists and is not under the VPA (`!containerAggregateState.IsUnderVPA`), and it's not stale (`!containerAggregateState.IsAggregateStale(now)`), then why are we keeping the recommendation instead of pruning it?

maxcao13 replied:
If the container no longer exists, then either:
* It's considered stale by grace period
* Not considered stale by grace period

If it is considered stale, then we are **not** going to overwrite the new recommendation with the previous, and we just ignore this branch of code which should set a new recommendation that we calculated in https://github.com/kubernetes/autoscaler/pull/6745/files#diff-5c44f22eb0f35931ab799838d2584dd10037916c74f114124e6904af48fcd608R96

If it's not considered stale, then potentially the user still want's this container's recommendation to exist for later, so here I explicitly overwrite the stale container's calculated recommendation with it's previous recommendation.

So the logic is sort of backwards, if that makes sense.

omerap12 replied:
Understood. Thanks for clarifying this :)

omerap12 reviewed (commented):
Looks good, I have a couple of notes.

maxcao13 commented:
> I have another comment, and I'm unsure on this one.
> 
> The new behaviour results in this:
> ...minimized
> 
> What I had originally expected to happen was that the `--pod-recommendation-min-memory-mb` would be split between the two containers, until the `vpaPruningGracePeriod` period expires, then the stale container would be removed, and `--pod-recommendation-min-memory-mb` be recalculated.
> 
> But, thinking it over, I guess this is fine? The second container's recommendations are retained (until it comes back, or when `vpaPruningGracePeriod` elapses). The primary container gets recommendations that make sense.

I agree I was unsure on what behaviour should be "correct".

On one hand, I think we want to keep the stale recommendation in there if the pruning hasn't happened, but here's a scenario that might be a problem.

1. We have 

```
    containerRecommendations:
    - containerName: hamster
      lowerBound:
        cpu: 445m
        memory: 262144k
      target:
        cpu: 627m
        memory: 262144k
      uncappedTarget:
        cpu: 627m
        memory: 262144k
      upperBound:
        cpu: "1"
        memory: 500Mi
    - containerName: hamster2
      lowerBound:
        cpu: 214m
        memory: 131072k
      target:
        cpu: 627m
        memory: 131072k
      uncappedTarget:
        cpu: 627m
        memory: 131072k
      upperBound:
        cpu: "1"
        memory: 500Mi
```
2. Then we restart the pod with the 2nd container added back.
3. I think the admission controller will set the pod container resources exactly like the recommendation above? Which is incorrect until the recommender loops once and recommends the correct resources.
I'm not exactly sure how we want to handle that. If we remove the recommendation entirely, that means the second container doesn't get a recommended resource initially at all, even if the pruning grace period wasn't met.

maxcao13 commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -422,6 +495,15 @@ func (feeder *clusterStateFeeder) LoadPods() {
+		if podExists && len(pod.Containers) > 1 {
+			feeder.clusterState.SetVPAContainersPerPod(podState, true)
+		} else if !podExists {
+			panic("This shouldn't happen because AddOrUpdatePod should've placed this pod in the clusterState")
```
Oh, sorry I forgot about that. That was a temporary check for me. I'll remove that in the next set of patches, thanks.

omerap12 commented:
> I have another comment, and I'm unsure on this one.
> 
> The new behaviour results in this:
> 
> ```
> status:
>   conditions:
>   - lastTransitionTime: "2025-01-11T12:07:41Z"
>     status: "True"
>     type: RecommendationProvided
>   recommendation:
>     containerRecommendations:
>     - containerName: hamster
>       lowerBound:
>         cpu: 445m
>         memory: 262144k
>       target:
>         cpu: 627m
>         memory: 262144k
>       uncappedTarget:
>         cpu: 627m
>         memory: 262144k
>       upperBound:
>         cpu: "1"
>         memory: 500Mi
>     - containerName: hamster2
>       lowerBound:
>         cpu: 214m
>         memory: 131072k
>       target:
>         cpu: 627m
>         memory: 131072k
>       uncappedTarget:
>         cpu: 627m
>         memory: 131072k
>       upperBound:
>         cpu: "1"
>         memory: 500Mi
> ```
> 
> This happens after the second container is removed, but before the `vpaPruningGracePeriod` period happens. The remaining container's memory is set to the value of `--pod-recommendation-min-memory-mb`, and the deleted container still exists.
> 
> What I had originally expected to happen was that the `--pod-recommendation-min-memory-mb` would be split between the two containers, until the `vpaPruningGracePeriod` period expires, then the stale container would be removed, and `--pod-recommendation-min-memory-mb` be recalculated.
> 
> But, thinking it over, I guess this is fine? The second container's recommendations are retained (until it comes back, or when `vpaPruningGracePeriod` elapses). The primary container gets recommendations that make sense.

For me it makes sense

adrianmoisey commented:
/milestone vertical-pod-autoscaler-1.4.0

adrianmoisey commented:
I assume the comment here can now be removed? https://github.com/kubernetes/autoscaler/blob/e179f01240a17d1dfb7487fba3468917b0942b04/vertical-pod-autoscaler/pkg/recommender/model/vpa.go#L94-L95

adrianmoisey commented:
Generally this seems fine. There are a few "TODO" items that I think need doing (such as the tests).
Pinging @voelzmo and @raywainman to get their eyes on it too

maxcao13 committed:
VPA: re-design aggregate pruning

Changes so that the recommender starts to prune stale aggregates only when all aggregates by containerName no longer belong to an active container. The aggregate of all containers of the same name are considered stale based on the "newest" aggregate state. i.e. the aggregate with the most recent update time.

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 committed:
VPA: Add unit tests for pruning grace period and containersPerPod field

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 commented:
In order, the most recent commits are:
* https://github.com/kubernetes/autoscaler/pull/6745/commits/3d4f7510d89d8457b199d8be8f3d380d222302f2 is a commit for e2e tests.
* https://github.com/kubernetes/autoscaler/pull/6745/commits/24c4eba97dd620abdfd0e4a079463beaeb82355d is a commit which reworks a bit of how the pruning works, in that if an aggregate container still exists, but is considered stale, it will get pruned unless there is a container with the same name that still exists. The grace period of the combined `aggregateByContainerName` will follow the aggregate with the most recent `updateTime`. I think this makes more sense as it follows how the VPA usually works with contributive container names. 
* https://github.com/kubernetes/autoscaler/pull/6745/commits/b8d78c6c757b6478a9fee69b540ca224a3963823 is a commit with unit tests.

maxcao13 committed:
fixup! VPA: re-design aggregate pruning

Signed-off-by: Max Cao <macao@redhat.com>

maxcao13 committed:
fixup! VPA: Add unit tests for pruning grace period and containersPerPod field

Signed-off-by: Max Cao <macao@redhat.com>

jkyros commented:
This was discussed in the 7/7/2025 sig-autoscaling meeting re: whether this was still wanted, sounded like the "momentum shift" was maybe due to in-place VPA taking precedence, and if we get this updated/rebased we can potentially keep this one alive :smile:
